### PR TITLE
Work around XAML Islands issues with TextCommandBarFlyout

### PR DESCRIPTION
**Pull request by rozele**
As reported in https://github.com/microsoft/microsoft-ui-xaml/issues/5341, the singleton TextCommandBarFlyout for the XAML TextBox and TextBlock components are limited to one window. When the flyout opens on a different window after already being opened on another, the first window the flyout was opened on gets focus, and the flyout is immediately closed. This change creates a TextCommandBarFlyout per TextInput native view and per selection TextBlock native view.

Additionally, it includes logic to work around an issue with the Proofing sub-menu flyout, which crashes due to an animation bug on Windows 10 versions lower than 21H1.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/8249)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 1]
```diff
@@ -804,6 +805,18 @@ ShadowNode *TextInputViewManager::createShadow() const {
 
 XamlView TextInputViewManager::CreateViewCore(int64_t /*tag*/, const winrt::Microsoft::ReactNative::JSValueObject &) {
   xaml::Controls::TextBox textBox;
+  // This works around a XAML Islands bug where the XamlRoot of the first
+  // window the flyout is shown on takes ownership of the flyout and attempts
+  // to show the flyout on other windows cause the first window to get focus.
+  // https://github.com/microsoft/microsoft-ui-xaml/issues/5341
+  if (IsXamlIsland() &&
+      winrt::Windows::Foundation::Metadata::ApiInformation::IsApiContractPresent(
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:813]: The contract check may be redundant, I don't think we can ever have the case that IsXamlIsland() == true if we are less that API level 8.

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 2]
```diff
@@ -804,6 +806,18 @@ ShadowNode *TextInputViewManager::createShadow() const {
...
+          L"Windows.Foundation.UniversalApiContract", 7)) {
+    xaml::Controls::TextCommandBarFlyout flyout;
+    flyout.Placement(xaml::Controls::Primitives::FlyoutPlacementMode::BottomEdgeAlignedLeft);
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:817]: should this be BottomEdgeAlignedRight for RTL?

File: vnext/Microsoft.ReactNative/Views/TextViewManager.cpp [hunk 3]
```diff
@@ -211,10 +213,24 @@ bool TextViewManager::UpdateProperty(
...
+      // to show the flyout on other windows cause the first window to get focus.
+      // https://github.com/microsoft/microsoft-ui-xaml/issues/5341
+      if (selectable && IsXamlIsland() &&
+          winrt::Windows::Foundation::Metadata::ApiInformation::IsApiContractPresent(
+              L"Windows.Foundation.UniversalApiContract", 7)) {
+        xaml::Controls::TextCommandBarFlyout flyout;
+        flyout.Placement(xaml::Controls::Primitives::FlyoutPlacementMode::BottomEdgeAlignedLeft);
+        textBlock.ContextFlyout(flyout);
+        textBlock.SelectionFlyout(flyout);
+      }
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/TextViewManager.cpp:230]: consider extracting this into a helper you can call from both places, like `EnsureTextFlyout<T>`

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp (hunk 1 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:814]: 19h1 is the earliest version with xaml islands support which is Api contract 10 IIRC so this should be safe to remove

**Review by asklar — APPROVED** [review]

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp (hunk 2 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:817]: Good question... let me check.

Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:817]: I set FlowDirection to RightToLeft and the Placement was still BottomEdgeAlignedLeft in a standalone UWP app. Anything else I should test?

Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:817]: I also set my system settings to Hebrew and the default was still BottomEdgeAlignedLeft. That being said, it looked like the app deployed in RTL mode (window Minimize/Maximize/Close buttons were on the top left), then when XAML rendered it switched back to LTR.

Comment (asklar) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:817]: thanks for checking : ) looks fine then!

File: vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj.filters [hunk 4]
```diff
@@ -218,9 +218,6 @@
     <ClCompile Include="Views\RawTextViewManager.cpp">
       <Filter>Views</Filter>
     </ClCompile>
-    <ClCompile Include="Views\ReactRootControl.cpp">
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Microsoft.ReactNative.vcxproj.filters:221]: Need to revert this change, not sure why it moved so much stuff.

File: vnext/Microsoft.ReactNative/Utils/XamlIslandUtils.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  // to show the flyout on other windows cause the first window to get focus.
+  // https://github.com/microsoft/microsoft-ui-xaml/issues/5341
+  if (IsXamlIsland()) {
+    xaml::Controls::TextCommandBarFlyout flyout;
+    flyout.Placement(xaml::Controls::Primitives::FlyoutPlacementMode::BottomEdgeAlignedLeft);
+
+    // This works around a XAML Islands bug where the Proofing sub-menu for
+    // TextBox causes a crash while animating to open / close before 21H1.
+    // https://github.com/microsoft/microsoft-ui-xaml/issues/3529
+    if (!Is21H1OrHigher() && textView.try_as<xaml::Controls::TextBox>()) {
```
Comment (rozele) [inline vnext/Microsoft.ReactNative/Utils/XamlIslandUtils.h:28]: Remove 21H1 check here, we have it in the method.

**Comment by rozele** [comment]
Easy way to repro these issues in playground-win32:
1. Comment out the code that creates a new thread for new windows in Playground-win32.cpp:
https://github.com/microsoft/react-native-windows/blob/main/packages/playground/windows/playground-win32/Playground-Win32.cpp#L133
2. For the multi-window issue, load two instances of RNTester and attempt to show the context menu on each window. Note that the first window will get focused when you attempt to show the context menu on the second.
3. For the Proofing menu crash, open RNTester, navigate to a TextInput field with autocorrect, type text that would have text suggestions, open the context menu and move the mouse around to open and close the proofing menu flyout multiple times on Windows 10 < 21H1.

**Comment by NickGerleman** [comment]
@rozele is this one ready for merge?

File: vnext/Microsoft.ReactNative/Views/TextViewManager.cpp [hunk 6]
```diff
@@ -211,10 +212,15 @@ bool TextViewManager::UpdateProperty(
...
-      textBlock.IsTextSelectionEnabled(propertyValue.AsBoolean());
-    else if (propertyValue.IsNull())
+    if (propertyValue.Type() == winrt::Microsoft::ReactNative::JSValueType::Boolean) {
+      const auto selectable = propertyValue.AsBoolean();
+      textBlock.IsTextSelectionEnabled(selectable);
+      if (selectable) {
+        EnsureUniqueTextFlyoutForXamlIsland(textBlock);
+      }
+    } else if (propertyValue.IsNull()) {
       textBlock.ClearValue(xaml::Controls::TextBlock::IsTextSelectionEnabledProperty());
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/Views/TextViewManager.cpp:222]: we should probably delete the custom type when clearing the value

**Comment by rozele** [comment]
@NickGerleman yes.

File: vnext/Microsoft.ReactNative/Views/TextViewManager.cpp (hunk 6 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextViewManager.cpp:222]: Sure, we can do that in this case.

Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextViewManager.cpp:222]: Done.

**Comment by rozele** [comment]
@NickGerleman yes.

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp [hunk 7]
```diff
@@ -804,6 +805,7 @@ ShadowNode *TextInputViewManager::createShadow() const {
 
 XamlView TextInputViewManager::CreateViewCore(int64_t /*tag*/, const winrt::Microsoft::ReactNative::JSValueObject &) {
   xaml::Controls::TextBox textBox;
+  EnsureUniqueTextFlyoutForXamlIsland(textBox);
```
Comment (lyahdav) [inline vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:808]: I wonder if this needs to be called in `registerEvents` since if the secureTextEntry 
 prop changes a new TextBox or PasswordBox is created here: https://github.com/microsoft/react-native-windows/blob/main/vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp#L508-L535

File: vnext/Microsoft.ReactNative/Views/TextViewManager.cpp (hunk 6 above)
Comment (lyahdav) [reply vnext/Microsoft.ReactNative/Views/TextViewManager.cpp:222]: Shouldn't we call `ClearUniqueTextFlyoutForXamlIsland` if selectable is set to false? Right now it's only called if it's set to null.

File: vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp (hunk 7 above)
Comment (rozele) [reply vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp:808]: Good call out.

**Comment by rozele** [comment]
@NickGerleman - going to implement @lyahdav's suggestion, then this should be ready 😅

**Comment by lyahdav** [comment]
@NickGerleman Any update on this? It's blocking another PR I'd like to submit.

**Comment by NickGerleman** [comment]
Sorry, missed the notification that it is ready for merge. I don't think GH updates send notifications, so feel free to give me a mention is there is a change to an older PR, and I will take a look.
